alibi_detect/od/pytorch/svm.py

```diff
@@ -0,0 +1,441 @@
+        """
+        if (isinstance(device, str) and device in ('gpu', 'cuda')) or \
+                (isinstance(device, torch.device) and device.type == 'cuda'):
+            warnings.warn(('The `sgd` optimization option is best suited for CPU. If '
+                           'you want to use GPU, consider using the `gd` option.'))
+
+        super().__init__(
+            device=device,
+            n_components=n_components,
+            kernel=kernel,
+            nu=nu,
+        )
+
+    def fit(  # type: ignore[override]
+        self,
+        x_ref: torch.Tensor,
+        tol: float = 1e-6,
+        max_iter: int = 1000,
+        verbose: int = 0,
+    ) -> Dict:
+        """Fit the SKLearn SVM model.
+
+        Parameters
+        ----------
+        x_ref
+            Training data.
+        tol
+            The decrease in loss required over the previous ``n_iter_no_change`` iterations in order to
+            continue optimizing.
+        max_iter
+            The maximum number of optimization steps.
+        verbose
+            Verbosity level during training. ``0`` is silent, ``1`` a progress bar.
+
+        Returns
+        -------
+        Dictionary with fit results. The dictionary contains the following keys:
+            - converged: `bool` indicating whether training converged.
+            - n_iter: number of iterations performed.
+        """
+        x_nys = self.nystroem.fit(x_ref).transform(x_ref)
+        self.svm = SGDOneClassSVM(
+            tol=tol,
+            max_iter=max_iter,
+            verbose=verbose,
+            nu=self.nu
+        )
+        x_nys = x_nys.cpu().numpy()
+        self.svm = self.svm.fit(x_nys)
+        self._set_fitted()
+        return {
+            'converged': self.svm.n_iter_ < max_iter,
+            'n_iter': self.svm.n_iter_,
+        }
+
+    def format_fit_kwargs(self, fit_kwargs: Dict) -> Dict:
+        """Format kwargs for `fit` method.
+
+        Parameters
+        ----------
+        fit_kwargs
+            dictionary of Kwargs to format. See `fit` method for details.
+
+        Returns
+        -------
+        Formatted kwargs.
+        """
+        return dict(
+            tol=fit_kwargs.get('tol', 1e-3),
+            max_iter=fit_kwargs.get('max_iter', 1000),
+            verbose=fit_kwargs.get('verbose', 0),
+        )
+
+    def score(self, x: torch.Tensor) -> torch.Tensor:
+        """Computes the score of `x`
+
+        Parameters
+        ----------
+        x
+            `torch.Tensor` with leading batch dimension.
+
+        Returns
+        -------
+        `torch.Tensor¬` of scores with leading batch dimension.
+
+        Raises
+        ------
+        NotFittedError
+            Raised if method called and detector has not been fit.
+        """
+        self.check_fitted()
+        x_nys = self.nystroem.transform(x)
+        x_nys = x_nys.cpu().numpy()
+        return self._to_tensor(-self.svm.score_samples(x_nys))
```

Comment:
Currently, due to differences between how we and `sklearn` set up our optimisations, scores returned via `sgd` vs `gd` variants differ by a linear transformation. Lets fix the intercept at 1 and scale coefficients accordingly such that we have consistency in this regard. (Note that out `intercept` vairable it equal to `sklearn`'s `offset` variable as `intercept = 1-offset`.)